### ISSUE TITLE
[gfx1201][Windows] BF16 training loss diverges from native Linux on the same RX 9070 XT (ROCm 7.2.1; reproduced with a 7.13 nightly wheel)

### ISSUE THREAD

**Issue by crivayne** [ROCm/TheRock#8095]
## Summary

On an RX 9070 XT (gfx1201, RDNA4), bf16 training with PyTorch converges systematically worse under the Windows-native ROCm stack than under a Linux ROCm stack on the exact same GPU. With hardware, model config, seed, and prepared data held constant — only the OS-specific PyTorch/ROCm software stack changed (Windows-native → Linux ROCm) — the final `val_bpb` improves from 1.9285 to 1.6439, and the Linux run also matches or beats an NVIDIA RTX 3060 Ti reference. The gap is roughly 10x the seed-to-seed standard deviation measured across three Windows-side seeds (i.e. not run-to-run noise), and none of the runtime knobs we tried (fp32 attention, full-reduction, Muon-fp32, BLAS-backend selection) removed it.

These results localize the discrepancy to the Windows-specific ROCm/PyTorch software stack; the responsible component or kernel has not yet been isolated. (The stable Windows and Linux runs use the same PyTorch release version string, but platform-specific PyTorch and ROCm code paths still differ, so this A/B does not by itself identify which component is responsible.)

## Symptom — same-GPU OS-stack A/B (bf16, eager, device_batch=8, seed=42, identical data + tokenizer)

Same RX 9070 XT, same stock `train.py`, same seed/batch — only the OS + ROCm software stack changes. Per-step training loss and final `val_bpb` (lower = better):

| step | Windows-native ROCm 7.2.1 | Linux ROCm (native Ubuntu), same GPU |
|---:|---:|---:|
| 0 | 9.0115 | 9.0115 |
| 5 | 8.0835 | 7.7294 |
| 10 | 7.2063 | 6.6874 |
| 20 | 6.3791 | 5.8603 |
| 40 | 5.8808 | 5.2675 |
| 60 | 5.6166 | 4.8501 |
| **final val_bpb** | **1.9285** | **1.6439** |

The two runs start at the identical step-0 loss (9.0115) — consistent with matched initialization and the same first batch (tokenizer hash and deterministic loader were also checked separately). They diverge by step 5, the deficit persists and generally grows at later steps, ending at a **0.285 `val_bpb` deficit** for the Windows-native stack. An earlier, very slow WSL2 (Linux ROCm) run on the same GPU showed the same direction (step 5: 7.692; step 9: 6.744), and an NVIDIA RTX 3060 Ti reference under the same recipe lands at step 5/10 = 8.166 / 7.022 — i.e. the AMD card on the Linux stack matches or beats the NVIDIA reference, while only the Windows-native stack trails.

**Not stale — reproduced on a 7.13 nightly.** Installing a gfx120X nightly wheel obtained from AMD's nightly index at test time (`torch 2.13.0a0+rocm7.13.0a20260416`, `torch.version.hip` 7.2.0) into a clean venv on the same GPU and rerunning the identical bench gives `val_bpb` 1.9582. That run completed fewer updates within the fixed 5-minute budget (65 steps / 34.1M tokens vs 70 / 36.7M), so its final `val_bpb` is partly throughput-sensitive; however its matched-step loss also trails Linux from step 10 onward, so the discrepancy remains reproducible on the newer build.

## What we ruled out (all on the Windows-native stack)

- **Seed noise** — three seeds (42/43/123) give val σ≈0.02, step-10 σ≈0.08; the same-GPU OS-stack gap is >10x that σ. Same-seed reruns match to 3 decimals (deterministic), so it is systematic.
- **Attention precision** — forcing SDPA to fp32: no change.
- **matmul reduced-precision reduction** — disabling bf16/fp16 reduced-precision reduction: no change (possibly a no-op on ROCm).
- **Muon optimizer orthogonalization in fp32** — no change.
- **Data / tokenizer / init** — tokenizer SHA-256 matches across machines, loader is deterministic (no RNG), init RNG produces the same sequence (only ~1e-7 last-bit differences).
- **hipBLASLt as sole culprit** — `preferred_blas_library("cublas")` (→ rocBLAS on ROCm) leaves the gap unchanged, so it is not hipBLASLt alone.
- **Full fp32 forward (autocast off), Windows only** — improved both the AMD and the NVIDIA runs, but the Windows AMD run still trailed the Windows NVIDIA reference at step 10 (7.09 vs 6.42). This is a cross-vendor comparison on Windows, not a same-GPU Windows/Linux fp32 A/B, so it does not by itself localize the OS-stack discrepancy; we have not yet run a same-GPU full-fp32 A/B.

Net: the tested runtime knobs did not remove the gap, and it reproduces on both the stable and a nightly Windows build; the responsible component or kernel has not yet been isolated.

### Operating System

Windows 10 (build 19045.6466), native — compared against Ubuntu 24.04 native (Linux ROCm), and WSL2 Ubuntu 24.04

### CPU

AMD Ryzen 9 7900X 12-Core Processor

### GPU

AMD Radeon RX 9070 XT (gfx1201, RDNA4), 16GB; NVIDIA RTX 3060 Ti used as a cross-vendor reference

### ROCm Version

7.2.1 (stable Windows) and a 7.13 nightly wheel (torch 2.13.0a0+rocm7.13.0a20260416; torch.version.hip 7.2.0) — both reproduce. Linux side: ROCm 7.2.1 native.

### Installation Method

Python wheels

### Installed ROCm Packages / Versions

Windows stable: `torch 2.9.1+rocm7.2.1` (ROCm 7.2.1 SDK wheels, repo.radeon.com rocm-rel-7.2.1).
Windows nightly: `torch 2.13.0a0+rocm7.13.0a20260416` with `rocm-sdk-core` / `rocm-sdk-libraries-gfx120X-all` 7.13.0a20260416, installed via `--index-url https://rocm.nightlies.amd.com/v2/gfx120X-all/ --pre torch`.
Linux: `torch 2.9.1+rocm7.2.1` (native Ubuntu 24.04).
Full `pip freeze` for either environment available on request.

### Steps to Reproduce

Public reproduction: https://github.com/crivayne/autoresearch-xv (fork of karpathy/autoresearch), tested at commit `5b0ec35`. Run once on Windows-native ROCm and once on Linux ROCm on the same card, identical settings.

Windows (PowerShell):
```powershell
python prepare.py
$env:AR_NO_COMPILE = "1"; $env:AR_SEED = "42"; $env:AR_DEVICE_BATCH_SIZE = "8"
python train.py
```

Linux (bash):
```bash
python prepare.py
AR_NO_COMPILE=1 AR_SEED=42 AR_DEVICE_BATCH_SIZE=8 python train.py
```

`prepare.py` fetches `karpathy/climbmix-400b-shuffle` shards + a BPE tokenizer (vocab 8192, fixed validation shard). Both runs start at the identical step-0 loss (9.0115); per-step loss diverges from step ~5 and final `val_bpb` differs: Windows-native 1.9285 (stable 7.2.1) / 1.9582 (7.13 nightly) vs Linux 1.6439. (`AR_DEVICE_BATCH_SIZE` defaults to 32, so passing 8 is required to match; `AR_NO_COMPILE=1` keeps both stacks in eager, since Windows has no Triton.)

### Additional Information

**Diagnostic data available (on request).** We currently have Windows-native bf16/fp32 one-step per-layer / per-parameter tensor-stat dumps (forward + backward). We can also collect matched Linux/NVIDIA dumps and Windows MIOpen + hipBLASLt logs using the settings from TheRock #2591 (`MIOPEN_ENABLE_LOGGING=1`, `MIOPEN_ENABLE_LOGGING_CMD=1`, `HIPBLASLT_LOG_MASK=32`, `HIPBLASLT_LOG_FILE=...`, `TORCH_BLAS_PREFER_HIPBLASLT=1`) if those would help triage.

**Questions for maintainers**
1. Is degraded bf16 training convergence on Windows-native ROCm for gfx1201 known / tracked? (Reproduced on both 7.2.1 stable and a 7.13 nightly wheel.)
2. Which logs would best localize it — MIOpen, hipBLASLt, or the per-layer forward/backward dump?
3. Is there a known-good kernel path (env/config) on Windows we should compare against?

**Related reports (not duplicates)**
- ROCm/TheRock #5284 (Windows gfx120X `torch._grouped_mm` null-kernel crash — a *crash*, fixed in a 7.13 wheel) — different symptom; this issue runs to completion but converges worse.
- ROCm/TheRock #2510 (RX 9070 XT Windows training *perf*), #2591 (gfx1201 MIOpen/hipBLASLt *perf* log collection) — performance, not convergence.
- ROCm/rocm-jax #84 (RDNA4 bf16 convolution *kernel selection* on WSL2).
- ROCm/ROCm #6034 (gfx1151 bf16 issues — different chip / OS / symptom).
